File: 0213-house-robber-ii/0213-house-robber-ii.rs
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.len() ==1{
            return nums[0];
        }
        let n = nums.len();
        let max1 =  Solution::rob_linear(&nums[0..n-1]);
        let max2 =  Solution::rob_linear(&nums[1..n]);
        max1.max(max2)

    }
    pub fn rob_linear(nums: &[i32]) -> i32{
        let mut prev1 =0;
        let mut prev2 =0;
        
        for &val in nums{
            let temp = prev1;
            prev1 = prev1.max(prev2 + val);
            prev2 = temp;
        }
        prev1
    }
}

File: 0213-house-robber-ii/0213-house-robber-ii.go
func rob(nums []int) int {
    n:= len(nums)
    if n == 0 {
        return 0
    }
    if n==1{
        return nums[0]
    }
    val1 := robLinear(nums[1:n])
    val2 := robLinear(nums[0:n-1])
    return max(val1, val2)

}

func robLinear(nums []int) int {
    prev1 := 0
    prev2 := 0
    for _, num := range nums{
        temp :=  prev1
        prev1 = max(prev1,prev2+num )
        prev2 =  temp
    } 
    return max(prev1, prev2)
}